Property "quarkus.class-loading.reloadable-artifacts" creates "ClassNotFoundException" in Arc

Update:
This property creates a ClassNotFoundException in arc when we wants to inject instances of CDI beans of this library.

Example: I want to inject all classes that implements a "NewServiceInterface" defined in the "libx" in my REST class of the app:

@Path("/hello")
public class GreetingResource {

    @Inject
    Instance<NewServiceInterface> services;
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        SingletonObject.getInstance().doSomething();
        SingletonObject2.getInstance().doSomething();
        System.out.println("services size: " +  services.stream().count());
        return "Hello RESTEasy again or nothing..";
    }

in the depedency "libx", I have this class NewServiceImpl, and the interface NewServiceInterface. 

@ApplicationScoped
public class NewServiceImpl implements NewServiceInterface {
}

Without quarkus.class-loading.reloadable-artifacts=org.acme.test:libx
=> No issue: I see "services size: 1"

With quarkus.class-loading.reloadable-artifacts=org.acme.test:libx
=> Class not found exception at first call of http://localhost:8080/hello (no reload at all)

2021-04-20 11:52:44,277 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /hello failed, error id: 69d2e706-5b85-49e5-b5c4-84c611c0f658-1: org.jboss.resteasy.spi.UnhandledException: java.lang.NoClassDefFoundError: org/acme/test/libx/NewServiceImpl
        at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
...
Caused by: java.lang.NoClassDefFoundError: org/acme/test/libx/NewServiceImpl
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:411)
...
Caused by: java.lang.ClassNotFoundException: org.acme.test.libx.NewServiceImpl                              
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)          
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)       
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)                                  
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:421)      
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:371)     

I tried adding
quarkus.class-loading.parent-first-artifacts=org.acme.test:libx
quarkus.arc.unremovable-types=org.acme.test.libx.**
Not working neither.

_Originally posted by @iriikoad in https://github.com/quarkusio/quarkus/issues/16633#issuecomment-823221087_

1-Compile le "libx" then "getting started" with mvn clean install
2-launch "getting-started" with mvn compile quarkus:dev
3-go to http://localhost:8080/hellocdi
You should have "2 services" displayed when it's working, but you'll get the exception
7d5-ac57-91ccb2c22bff-1: org.jboss.resteasy.spi.UnhandledException: java.lang.NoClassDefFoundError: org/acme/test/libx/NewServiceImpl
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:41)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:832)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.NoClassDefFoundError: org/acme/test/libx/NewServiceImpl
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:411)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:371)
	at org.acme.test.libx.NewServiceImpl_Bean.proxy(NewServiceImpl_Bean.zig:36)
	at org.acme.test.libx.NewServiceImpl_Bean.get(NewServiceImpl_Bean.zig:198)
	at org.acme.test.libx.NewServiceImpl_Bean.get(NewServiceImpl_Bean.zig:214)
	at io.quarkus.arc.impl.InstanceImpl.getBeanInstance(InstanceImpl.java:210)
	at io.quarkus.arc.impl.InstanceImpl.access$200(InstanceImpl.java:32)
	at io.quarkus.arc.impl.InstanceImpl$InstanceIterator.next(InstanceImpl.java:246)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:257)
	at java.base/java.util.stream.ReduceOps$5.evaluateSequential(ReduceOps.java:248)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.count(ReferencePipeline.java:605)
	at org.acme.getting.started.GreetingCdiResource.hello(GreetingCdiResource.java:27)
	at org.acme.getting.started.GreetingCdiResource_Subclass.hello$$superaccessor1(GreetingCdiResource_Subclass.zig:189)
	at org.acme.getting.started.GreetingCdiResource_Subclass$$function$$1.apply(GreetingCdiResource_Subclass$$function$$1.zig:29)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:63)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.getting.started.GreetingCdiResource_Subclass.hello(GreetingCdiResource_Subclass.zig:147)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	... 18 more
Caused by: java.lang.ClassNotFoundException: org.acme.test.libx.NewServiceImpl
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:421)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:371)
	... 63 more



[bug-arc-class-reloadable.zip](https://github.com/quarkusio/quarkus/files/6356886/bug-arc-class-reloadable.zip)

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
@@ -178,6 +178,9 @@ public class CuratedApplication implements Serializable, AutoCloseable {
             //any of the runtime artifacts, or user classes
             //this will load any deployment artifacts from the parent CL if they are present
             for (AppDependency i : appModel.getFullDeploymentDeps()) {
+                if (configuredClassLoading.reloadableArtifacts.contains(i.getArtifact().getKey())) {
+                    continue;
+                }
                 processCpElement(i.getArtifact(), element -> addCpElement(builder, i.getArtifact(), element));
             }
 
@@ -269,6 +272,11 @@ public class CuratedApplication implements Serializable, AutoCloseable {
         for (AdditionalDependency i : quarkusBootstrap.getAdditionalApplicationArchives()) {
             for (Path root : i.getArchivePath()) {
                 builder.addElement(ClassPathElement.fromPath(root));
+            }
+        }
+        for (AppDependency dependency : appModel.getUserDependencies()) {
+            if (configuredClassLoading.reloadableArtifacts.contains(dependency.getArtifact().getKey())) {
+                processCpElement(dependency.getArtifact(), element -> addCpElement(builder, dependency.getArtifact(), element));
             }
         }
         return builder.build();

